README.md
# Blind Index

Securely search encrypted database fields

Designed for use with [attr_encrypted](https://github.com/attr-encrypted/attr_encrypted)

Here’s a [full example](https://ankane.org/securing-user-emails-in-rails) of how to use it

Check out [this post](https://ankane.org/sensitive-data-rails) for more info on securing sensitive data with Rails

[![Build Status](https://travis-ci.org/ankane/blind_index.svg?branch=master)](https://travis-ci.org/ankane/blind_index)

## How It Works

We use [this approach](https://paragonie.com/blog/2017/05/building-searchable-encrypted-databases-with-php-and-sql) by Scott Arciszewski. To summarize, we compute a keyed hash of the sensitive data and store it in a column. To query, we apply the keyed hash function to the value we’re searching and then perform a database search. This results in performant queries for exact matches. `LIKE` queries are not possible, but you can index expressions.

## Leakage

An important consideration in searchable encryption is leakage, which is information an attacker can gain. Blind indexing leaks that rows have the same value. If you use this for a field like last name, an attacker can use frequency analysis to predict the values. In an active attack where an attacker can control the input values, they can learn which other values in the database match.

Here’s a [great article](https://blog.cryptographyengineering.com/2019/02/11/attack-of-the-week-searchable-encryption-and-the-ever-expanding-leakage-function/) on leakage in searchable encryption. Blind indexing has the same leakage as deterministic encryption.

## Installation

Add this line to your application’s Gemfile:

```ruby
gem 'argon2', '>= 2'
gem 'blind_index'
```

## Getting Started

> Note: Your model should already be set up with attr_encrypted. The examples are for a `User` model with `attr_encrypted :email`. See the [full example](https://ankane.org/securing-user-emails-in-rails) if needed.

Create a migration to add a column for the blind index

```ruby
add_column :users, :email_bidx, :string
add_index :users, :email_bidx
```

Next, generate a key

```ruby
SecureRandom.hex(32)
```

Store the key with your other secrets. This is typically Rails credentials or an environment variable ([dotenv](https://github.com/bkeepers/dotenv) is great for this). Be sure to use different keys in development and production, and be sure this is different than the key you use for encryption. Keys don’t need to be hex-encoded, but it’s often easier to store them this way.

Here’s a key you can use in development

```sh
USER_EMAIL_BLIND_INDEX_KEY=ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
```

Add to your model

```ruby
class User < ApplicationRecord
  blind_index :email, key: ENV["USER_EMAIL_BLIND_INDEX_KEY"]
end
```

Backfill existing records

```ruby
User.find_each do |user|
  user.compute_email_bidx
  user.save!
end
```

And query away

```ruby
User.where(email: "test@example.org")
```

## Validations

To prevent duplicates, use:

```ruby
class User < ApplicationRecord
  validates :email, uniqueness: true
end
```

We also recommend adding a unique index to the blind index column through a database migration.

## Expressions

You can apply expressions to attributes before indexing and searching. This gives you the the ability to perform case-insensitive searches and more.

```ruby
class User < ApplicationRecord
  blind_index :email, expression: ->(v) { v.downcase } ...
end
```

## Multiple Indexes

You may want multiple blind indexes for an attribute. To do this, add another column:

```ruby
add_column :users, :email_ci_bidx, :string
add_index :users, :email_ci_bidx
```

Update your model

```ruby
class User < ApplicationRecord
  blind_index :email, ...
  blind_index :email_ci, attribute: :email, expression: ->(v) { v.downcase } ...
end
```

Backfill existing records

```ruby
User.find_each do |user|
  user.compute_email_ci_bidx
  user.save!
end
```

And query away

```ruby
User.where(email_ci: "test@example.org")
```

## Index Only

If you don’t need to store the original value (for instance, when just checking duplicates), use a virtual attribute:

```ruby
class User < ApplicationRecord
  attribute :email
  blind_index :email, ...
end
```

*Requires ActiveRecord 5.1+*

## Multiple Columns

You can also use virtual attributes to index data from multiple columns:

```ruby
class User < ApplicationRecord
  attribute :initials

  # must come before the blind_index method so it runs first
  before_validation :set_initials, if: -> { changes.key?(:first_name) || changes.key?(:last_name) }

  blind_index :initials, ...

  def set_initials
    self.initials = "#{first_name[0]}#{last_name[0]}"
  end
end
```

*Requires ActiveRecord 5.1+*

## Algorithm

Argon2id is used for best security. The default cost parameters are 4 iterations and 32 MB memory.

Blind indexes are expensive to compute by default. For less sensitive fields, you can speed up computation with:

```ruby
class User < ApplicationRecord
  blind_index :email, fast: true, ...
end
```

This uses 3 iterations and 4 MB of memory.

> Note: This requires you to recompute the blind index

A number of other algorithms are [also supported](docs/Other-Algorithms.md). Unless you have specific reasons to use them, go with Argon2id.

## Key Rotation

To rotate keys without downtime, add a new column:

```ruby
add_column :users, :email_v2_bidx, :string
add_index :users, :email_v2_bidx
```

And add to your model

```ruby
class User < ApplicationRecord
  blind_index :email, key: ENV["USER_EMAIL_BLIND_INDEX_KEY"]
  blind_index :email_v2, attribute: :email, key: ENV["USER_EMAIL_V2_BLIND_INDEX_KEY"]
end
```

Backfill the data

```ruby
User.find_each do |user|
  user.compute_email_v2_bidx
  user.save!
end
```

Then update your model

```ruby
class User < ApplicationRecord
  blind_index :email, bidx_attribute: :encrypted_email_v2_bidx, key: ENV["EMAIL_V2_BLIND_INDEX_KEY"]

  # remove this line after dropping column
  self.ignored_columns = ["email_bidx"]
end
```

Finally, drop the old column.

## Fixtures

You can use encrypted attributes and blind indexes in fixtures with:

```yml
test_user:
  encrypted_email: <%= User.encrypt_email("test@example.org", iv: Base64.decode64("0000000000000000")) %>
  encrypted_email_iv: "0000000000000000"
  email_bidx: <%= User.compute_email_bidx("test@example.org").inspect %>
```

Be sure to include the `inspect` at the end, or it won’t be encoded properly in YAML.

## Reference

By default, blind indexes are encoded in Base64. Set a different encoding with:

```ruby
class User < ApplicationRecord
  blind_index :email, encode: ->(v) { [v].pack("H*") }
end
```

By default, blind indexes are 32 bytes. Set a smaller size with:

```ruby
class User < ApplicationRecord
  blind_index :email, size: 16
end
```

## Alternatives

One alternative to blind indexing is to use a deterministic encryption scheme, like [AES-SIV](https://github.com/miscreant/miscreant). In this approach, the encrypted data will be the same for matches.

## Upgrading

### 1.0.0

For existing fields, add:

```ruby
class User < ApplicationRecord
  blind_index :email, legacy: true, ...
end
```

#### Optional

To rotate to new fields, add a new column without the `encrypted_` prefix:

```ruby
add_column :users, :email_bidx, :string
add_index :users, :email_bidx
```

And add to your model

```ruby
class User < ApplicationRecord
  blind_index :email, key: ENV["USER_EMAIL_BLIND_INDEX_KEY"], legacy: true
  blind_index :email_upgraded, attribute: :email, bidx_attribute: :email_bidx, key: ENV["USER_EMAIL_BLIND_INDEX_KEY"]
end
```

Backfill the data

```ruby
User.find_each do |user|
  user.compute_email_upgraded_bidx
  user.save!
end
```

Then update your model

```ruby
class User < ApplicationRecord
  blind_index :email, key: ENV["USER_EMAIL_BLIND_INDEX_KEY"]

  # remove this line after dropping column
  self.ignored_columns = ["encrypted_email_bidx"]
end
```

Finally, drop the old column.

### 0.3.0

This version introduces a breaking change to enforce secure key generation. An error is thrown if your blind index key isn’t both binary and 32 bytes.

We recommend rotating your key if it doesn’t meet this criteria. You can generate a new key in the Rails console with:

```ruby
SecureRandom.hex(32)
```

Update your model to convert the hex key to binary.

```ruby
class User < ApplicationRecord
  blind_index :email, key: [ENV["USER_EMAIL_BLIND_INDEX_KEY"]].pack("H*")
end
```

And recompute the blind index.

```ruby
User.find_each do |user|
  user.compute_email_bidx
  user.save!
end
```

To continue without rotating, set:

```ruby
class User < ApplicationRecord
  blind_index :email, insecure_key: true, ...
end
```

## History

View the [changelog](https://github.com/ankane/blind_index/blob/master/CHANGELOG.md)

## Contributing

Everyone is encouraged to help improve this project. Here are a few ways you can help:

- [Report bugs](https://github.com/ankane/blind_index/issues)
- Fix bugs and [submit pull requests](https://github.com/ankane/blind_index/pulls)
- Write, clarify, or fix documentation
- Suggest or add new features

To get started with development and testing:

```sh
git clone https://github.com/ankane/blind_index.git
cd blind_index
bundle install
rake test
```


CHANGELOG.md
## 0.3.5 [unreleased]

- Added support for hex keys
- Added support for Argon2id
- Fixed querying with array values

## 0.3.4

- Added `size` option
- Added sanity checks for Argon2 cost parameters
- Fixed ActiveRecord callback issues introduced in 0.3.3

## 0.3.3

- Added support for string keys in finders

## 0.3.2

- Added support for dynamic finders
- Added support for inherited models

## 0.3.1

- Added scrypt and Argon2 algorithms
- Added `cost` option

## 0.3.0

- Enforce secure key generation
- Added `encode` option
- Added `default_options` method

## 0.2.1

- Added class method to compute blind index
- Fixed issue with cached statements

## 0.2.0

- Added support for ActiveRecord 4.2
- Improved validation support when multiple blind indexes
- Fixed `nil` handling

## 0.1.1

- Added support for ActiveRecord 5.2
- Added `callback` option
- Added support for `key` proc
- Fixed error inheritance

## 0.1.0

- First release


test/test_helper.rb
require "bundler/setup"
require "active_record"
require "attr_encrypted"
Bundler.require(:default)
require "minitest/autorun"
require "minitest/pride"
require "scrypt"
require "argon2"

ActiveRecord::Base.establish_connection adapter: "sqlite3", database: ":memory:"

if ENV["VERBOSE"]
  ActiveRecord::Base.logger = ActiveSupport::Logger.new(STDOUT)
end

ActiveRecord::Migration.create_table :users do |t|
  t.string :encrypted_email
  t.string :encrypted_email_iv
  t.string :email_bidx
  t.string :email_ci_bidx
  t.binary :email_binary_bidx
  t.string :encrypted_first_name
  t.string :encrypted_first_name_iv
  t.string :encrypted_last_name
  t.string :encrypted_last_name_iv
  t.string :initials_bidx
end

class User < ActiveRecord::Base
  attribute :initials, ActiveRecord::Type::String.new if ActiveRecord::VERSION::MAJOR >= 5

  before_validation :set_initials, if: -> { changes.key?(:first_name) || changes.key?(:last_name) }

  attr_encrypted :email, key: SecureRandom.random_bytes(32)
  attr_encrypted :first_name, key: SecureRandom.random_bytes(32)
  attr_encrypted :last_name, key: SecureRandom.random_bytes(32)

  blind_index :email, key: SecureRandom.random_bytes(32)
  blind_index :email_ci, algorithm: :scrypt, key: SecureRandom.random_bytes(32), attribute: :email, expression: ->(v) { v.try(:downcase) }, fast: true
  blind_index :email_binary, algorithm: :argon2, key: SecureRandom.random_bytes(32), attribute: :email, encode: false
  blind_index :initials, key: SecureRandom.hex(32), algorithm: :argon2id, size: 16

  validates :email, uniqueness: true
  validates :email_ci, uniqueness: true

  def set_initials
    self.initials = [first_name.first, last_name.first].join
  end

  # ensure custom method still works
  def read_attribute_for_validation(key)
    super
  end
end

class ActiveUser < User
  blind_index :child, key: SecureRandom.random_bytes(32)
end


test/blind_index_test.rb
require_relative "test_helper"

class BlindIndexTest < Minitest::Test
  def setup
    User.delete_all
  end

  def test_find_by
    create_user
    assert User.find_by(email: "test@example.org")
  end

  def test_find_by_string_key
    create_user
    assert User.find_by({"email" => "test@example.org"})
  end

  def test_dynamic_finders
    user = create_user
    assert User.find_by_email("test@example.org")
    assert User.find_by_id_and_email(user.id, "test@example.org")
  end

  def test_where
    create_user
    assert User.where(email: "test@example.org").first
  end

  def test_where_array
    create_user
    create_user(email: "test2@example.org")
    assert_equal 2, User.where(email: ["test@example.org", "test2@example.org"]).count
  end

  def test_where_string_key
    create_user
    assert User.where({"email" => "test@example.org"}).first
  end

  def test_where_not
    create_user
    assert User.where.not(email: "test2@example.org").first
  end

  def test_where_not_empty
    create_user
    assert_nil User.where.not(email: "test@example.org").first
  end

  def test_expression
    create_user
    assert User.where(email_ci: "TEST@example.org").first
  end

  def test_expression_different_case
    create_user email: "TEST@example.org"
    assert User.where(email_ci: "test@example.org").first
  end

  def test_encode
    user = create_user
    assert User.find_by(email_binary: "test@example.org")
    assert_equal 32, user.email_binary_bidx.bytesize
  end

  def test_validation
    create_user
    user = User.new(email: "test@example.org")
    assert !user.valid?
    assert_equal "Email has already been taken", user.errors.full_messages.first
  end

  def test_validation_case_insensitive
    create_user
    user = User.new(email: "TEST@example.org")
    assert !user.valid?
    assert_equal "Email ci has already been taken", user.errors.full_messages.first
  end

  def test_nil
    user = create_user(email: nil)
    assert_nil user.email_bidx
    assert User.where(email: nil).first
    assert_nil User.where(email: "").first
  end

  def test_empty_string
    user = create_user(email: "")
    assert user.email_bidx
    assert User.where(email: "").first
    assert_nil User.where(email: nil).first
  end

  def test_unset
    user = create_user
    user.email = nil
    user.save!
    assert_nil user.email_bidx
    assert User.where(email: nil).first
  end

  def test_class_method
    user = create_user
    assert_equal user.email_bidx, User.compute_email_bidx("test@example.org")
  end

  def test_secure_key
    error = assert_raises(BlindIndex::Error) do
      BlindIndex.generate_bidx("test@example.org", key: "bad")
    end
    assert_equal "Key must use binary encoding", error.message
  end

  # def test_secure_key_ascii
  #   error = assert_raises(BlindIndex::Error) do
  #     BlindIndex.generate_bidx("test@example.org", key: ("0"*32).encode("BINARY"))
  #   end
  #   assert_equal "Key must not be ASCII", error.message
  # end

  def test_secure_key_length
    error = assert_raises(BlindIndex::Error) do
      BlindIndex.generate_bidx("test@example.org", key: SecureRandom.random_bytes(20))
    end
    assert_equal "Key must be 32 bytes", error.message
  end

  def test_inheritance
    assert_equal %i[email email_ci email_binary initials], User.blind_indexes.keys
    assert_equal %i[email email_ci email_binary initials child], ActiveUser.blind_indexes.keys
  end

  def test_initials
    skip unless ActiveRecord::VERSION::MAJOR >= 5

    user = create_user(first_name: "Test", last_name: "User")
    assert User.find_by(initials: "TU")

    user = User.last
    user.email = "test2@example.org"
    user.save!
    assert User.find_by(initials: "TU")
  end

  def test_size
    result = BlindIndex.generate_bidx("secret", key: random_key, size: 16, encode: false)
    assert_equal 16, result.bytesize
  end

  def test_invalid_size
    error = assert_raises(BlindIndex::Error) do
      BlindIndex.generate_bidx("secret", key: random_key, size: 0, encode: false)
    end
    assert_equal "Size must be between 1 and 32", error.message
  end

  def text_backfill
    create_user
    User.update_all(email_bidx: nil)
    user = User.last
    assert_nil user.email_bidx
    user.compute_email_bidx
    assert user.email_bidx
  end

  private

  def random_key
    SecureRandom.random_bytes(32)
  end

  def create_user(email: "test@example.org", **attributes)
    User.create!({email: email}.merge(attributes))
  end
end


test/ci/before_install.sh
#!/bin/sh

set -e

LIBSODIUM_VERSION="1.0.16"

if [ ! -d "$HOME/libsodium" ]; then
  wget "https://download.libsodium.org/libsodium/releases/libsodium-$LIBSODIUM_VERSION.tar.gz"
  tar xvfz "libsodium-$LIBSODIUM_VERSION.tar.gz"
  cd "libsodium-$LIBSODIUM_VERSION"
  ./configure --prefix=$HOME/libsodium
  make && make check
  make install
fi


lib/blind_index/model.rb
module BlindIndex
  module Model
    def blind_index(name, key: nil, iterations: nil, attribute: nil, expression: nil, bidx_attribute: nil, callback: true, algorithm: nil, insecure_key: nil, encode: nil, cost: nil, size: nil, legacy: false, fast: false)
      iterations ||= 10000
      attribute ||= name
      bidx_attribute ||= legacy ? :"encrypted_#{name}_bidx" : :"#{name}_bidx"

      name = name.to_sym
      attribute = attribute.to_sym
      method_name = :"compute_#{name}_bidx"

      class_eval do
        @blind_indexes ||= {}

        unless respond_to?(:blind_indexes)
          def self.blind_indexes
            parent_indexes =
              if superclass.respond_to?(:blind_indexes)
                superclass.blind_indexes
              else
                {}
              end

            parent_indexes.merge(@blind_indexes || {})
          end
        end

        raise BlindIndex::Error, "Duplicate blind index: #{name}" if blind_indexes[name]

        @blind_indexes[name] = {
          key: key,
          iterations: iterations,
          attribute: attribute,
          expression: expression,
          bidx_attribute: bidx_attribute,
          algorithm: algorithm,
          insecure_key: insecure_key,
          encode: encode,
          cost: cost,
          size: size,
          legacy: legacy,
          fast: fast
        }.reject { |_, v| v.nil? }

        # should have been named generate_#{name}_bidx
        define_singleton_method method_name do |value|
          BlindIndex.generate_bidx(value, blind_indexes[name])
        end

        define_method method_name do
          self.send("#{bidx_attribute}=", self.class.send(method_name, send(attribute)))
        end

        if callback
          before_validation method_name, if: -> { changes.key?(attribute.to_s) }
        end

        # use include so user can override
        include InstanceMethods if blind_indexes.size == 1
      end
    end
  end

  module InstanceMethods
    def read_attribute_for_validation(key)
      if (bi = self.class.blind_indexes[key])
        send(bi[:attribute])
      else
        super
      end
    end
  end
end


lib/blind_index.rb
# dependencies
require "active_support"

# modules
require "blind_index/model"
require "blind_index/version"

module BlindIndex
  class Error < StandardError; end

  class << self
    attr_accessor :default_options
  end
  self.default_options = {
    encode: true,
    cost: {}
  }

  def self.generate_bidx(value, key:, **options)
    options = default_options.merge(options)

    # apply expression
    value = options[:expression].call(value) if options[:expression]

    unless value.nil?
      algorithm = (options[:algorithm] || (options[:legacy] ? :pbkdf2_sha256 : :argon2id)).to_sym
      algorithm = :pbkdf2_sha256 if algorithm == :pbkdf2_hmac
      algorithm = :argon2i if algorithm == :argon2

      key = key.call if key.respond_to?(:call)
      raise BlindIndex::Error, "Missing key for blind index" unless key

      key = key.to_s
      unless options[:insecure_key] && algorithm == :pbkdf2_sha256
        # decode hex key
        if key.encoding != Encoding::BINARY && key =~ /\A[0-9a-f]{64}\z/i
          key = [key].pack("H*")
        end

        raise BlindIndex::Error, "Key must use binary encoding" if key.encoding != Encoding::BINARY
        raise BlindIndex::Error, "Key must be 32 bytes" if key.bytesize != 32
      end

      # gist to compare algorithm results
      # https://gist.github.com/ankane/fe3ac63fbf1c4550ee12554c664d2b8c
      cost_options = options[:cost]

      # check size
      size = (options[:size] || 32).to_i
      raise BlindIndex::Error, "Size must be between 1 and 32" unless (1..32).include?(size)

      value = value.to_s

      value =
        case algorithm
        when :argon2id
          t = (cost_options[:t] || (options[:fast] ? 3 : 4)).to_i
          # use same bounds as rbnacl
          raise BlindIndex::Error, "t must be between 3 and 10" if t < 3 || t > 10

          # m is memory in kibibytes (1024 bytes)
          m = (cost_options[:m] || (options[:fast] ? 12 : 15)).to_i
          # use same bounds as rbnacl
          raise BlindIndex::Error, "m must be between 3 and 22" if m < 3 || m > 22

          [Argon2::Engine.hash_argon2id(value, key, t, m, size)].pack("H*")
        when :pbkdf2_sha256
          iterations = cost_options[:iterations] || options[:iterations] || 10000
          OpenSSL::PKCS5.pbkdf2_hmac(value, key, iterations, size, "sha256")
        when :argon2i
          t = (cost_options[:t] || 3).to_i
          # use same bounds as rbnacl
          raise BlindIndex::Error, "t must be between 3 and 10" if t < 3 || t > 10

          # m is memory in kibibytes (1024 bytes)
          m = (cost_options[:m] || 12).to_i
          # use same bounds as rbnacl
          raise BlindIndex::Error, "m must be between 3 and 22" if m < 3 || m > 22

          # 32 byte digest size was limitation of argon2 gem before 2.0
          # we could conditionally check, but argon2id is now recommended
          raise BlindIndex::Error, "Size must be 32" unless size == 32
          [Argon2::Engine.hash_argon2i(value, key, t, m)].pack("H*")
        when :scrypt
          n = cost_options[:n] || 4096
          r = cost_options[:r] || 8
          cp = cost_options[:p] || 1
          SCrypt::Engine.scrypt(value, key, n, r, cp, size)
        else
          raise BlindIndex::Error, "Unknown algorithm"
        end

      encode = options[:encode]
      if encode
        if encode.respond_to?(:call)
          encode.call(value)
        else
          [value].pack(options[:legacy] ? "m" : "m0")
        end
      else
        value
      end
    end
  end
end

ActiveSupport.on_load(:active_record) do
  require "blind_index/extensions"
  extend BlindIndex::Model

  if defined?(ActiveRecord::TableMetadata)
    ActiveRecord::TableMetadata.prepend(BlindIndex::Extensions::TableMetadata)
  else
    ActiveRecord::PredicateBuilder.singleton_class.prepend(BlindIndex::Extensions::PredicateBuilder)
  end

  ActiveRecord::DynamicMatchers::Method.prepend(BlindIndex::Extensions::DynamicMatchers)

  unless ActiveRecord::VERSION::STRING.start_with?("5.1.")
    ActiveRecord::Validations::UniquenessValidator.prepend(BlindIndex::Extensions::UniquenessValidator)
  end
end
